merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/SimulationDriver.java

```diff
@@ -90,8 +111,223 @@ SimulationResults simulate(
       }
 
-      return SimulationEngine.computeResults(engine, startTime, elapsedTime, activityTopic, timeline, missionModel);
+      return computeResults(engine, startTime, elapsedTime, activityTopic, timeline, missionModel);
+    }
+  }
+
+  /** Compute a set of results from the current state of simulation. */
+  // TODO: Move result extraction out of the SimulationEngine.
+  //   The Engine should only need to stream events of interest to a downstream consumer.
+  //   The Engine cannot be cognizant of all downstream needs.
+  // TODO: Whatever mechanism replaces `computeResults` also ought to replace `isTaskComplete`.
+  // TODO: Produce results for all tasks, not just those that have completed.
+  //   Planners need to be aware of failed or unfinished tasks.
+  public static SimulationResults computeResults(
+      final SimulationEngine engine,
+      final Instant startTime,
+      final Duration elapsedTime,
+      final Topic<ActivityInstanceId> activityTopic,
+      final TemporalEventSource timeline,
+      final MissionModel<?> missionModel
+  ) {
+
+    // Collect per-task information from the event graph.
+    final var taskInfo = new TaskInfo();
+
+    for (final var point : timeline) {
+      if (!(point instanceof TemporalEventSource.TimePoint.Commit p)) continue;
+
+      final var trait = new TaskInfo.Trait(missionModel, activityTopic);
+      p.events().evaluate(trait, trait::atom).accept(taskInfo);
+    }
+
+    // Extract profiles for every resource.
+    final var realProfiles = new HashMap<String, List<Pair<Duration, RealDynamics>>>();
+    final var discreteProfiles = new HashMap<String, Pair<ValueSchema, List<Pair<Duration, SerializedValue>>>>();
+
+    engine.getResources().forEach((id, state) -> {
+      final var name = id.id();
+      final var resource = state.resource();
+
+      switch (resource.getType()) {
+        case "real" -> realProfiles.put(
+            name,
+            serializeProfile(elapsedTime, state, SimulationDriver::extractRealDynamics));
+
+        case "discrete" -> discreteProfiles.put(
+            name,
+            Pair.of(
+                state.resource().getSchema(),
+                serializeProfile(elapsedTime, state, Resource::serialize)));
+
+        default ->
+            throw new IllegalArgumentException(
+                "Resource `%s` has unknown type `%s`".formatted(name, resource.getType()));
+      }
+    });
+
+    // Give every task corresponding to a child activity an ID that doesn't conflict with any root activity.
+    final var taskToPlannedDirective = new HashMap<>(taskInfo.taskToPlannedDirective());
+    final var usedActivityInstanceIds =
+        taskToPlannedDirective
+            .values()
+            .stream()
+            .map(ActivityInstanceId::id)
+            .collect(Collectors.toSet());
+
+    // Grab all tasks that are activities, so we don't have to check later
+    final var activityTasks =
+        engine
+            .getTasks()
```

Comment:
One refactor that I tried on my end was to add a `getActivityTasks` method to `TaskInfo` (which returns a `Set<TaskId>`), and then have the engine provide a `getTaskState(TaskId)` method, but upon reflection, I don't know if that gains us anything.